modules/dlm/src/test/java/org/elasticsearch/dlm/DataLifecycleServiceTests.java

```diff
@@ -0,0 +1,235 @@
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.junit.After;
+import org.junit.Before;
+import org.mockito.ArgumentCaptor;
+import org.mockito.Mockito;
+
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+
+import static org.elasticsearch.cluster.metadata.DataStreamTestHelper.newInstance;
+import static org.hamcrest.Matchers.is;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class DataLifecycleServiceTests extends ESTestCase {
+
+    private ClusterService clusterService;
+    private long now;
+    private ThreadPool threadPool;
+    private IndicesAdminClient indicesClient;
+    private DataLifecycleService dataLifecycleService;
+    ArgumentCaptor<RolloverRequest> rolloverRequestCaptor;
+    ArgumentCaptor<DeleteIndexRequest> deleteRequestCaptor;
+
+    @Before
+    public void setupServices() {
+        clusterService = mock(ClusterService.class);
+        now = randomNonNegativeLong();
+        Clock clock = Clock.fixed(Instant.ofEpochMilli(now), ZoneId.of(randomFrom(ZoneId.getAvailableZoneIds())));
+
+        doAnswer(invocationOnMock -> null).when(clusterService).addListener(any());
+
+        Settings settings = Settings.builder().put(DataLifecycleService.DLM_POLL_INTERVAL, "1s").build();
+        when(clusterService.getClusterSettings()).thenReturn(
+            new ClusterSettings(settings, Collections.singleton(DataLifecycleService.DLM_POLL_INTERVAL_SETTING))
+        );
+        when(clusterService.lifecycleState()).thenReturn(Lifecycle.State.STARTED);
+
+        Client client = mock(Client.class);
+        AdminClient adminClient = mock(AdminClient.class);
+        indicesClient = mock(IndicesAdminClient.class);
+        when(client.admin()).thenReturn(adminClient);
+        when(adminClient.indices()).thenReturn(indicesClient);
+        when(client.settings()).thenReturn(Settings.EMPTY);
+
+        threadPool = new TestThreadPool(getTestName());
+        dataLifecycleService = new DataLifecycleService(Settings.EMPTY, client, clusterService, clock, threadPool, () -> now);
+        dataLifecycleService.init();
+        Mockito.verify(clusterService).addListener(dataLifecycleService);
+
+        rolloverRequestCaptor = ArgumentCaptor.forClass(RolloverRequest.class);
+        deleteRequestCaptor = ArgumentCaptor.forClass(DeleteIndexRequest.class);
+    }
```

Comment:
I know it's a huge pain, but I think we should endeavor to avoid mocking for these, I know a large part of them are two things — cluster service and a client.

For `ClusterService` we now have `ClusterServiceUtils` which will let us create a real ClusterService that isn't mocked, and then call `setState(...)` on it.

For the client we can create a new class subclassing `NoOpClient` that will allow us to capture/fake whatever requests or responses we need.

I'm hoping by doing that we end up with less brittle testing that's closer to real-life usage.

Reply:
Thanks for the suggestion  - both mocks have been removed 🎉